### PR TITLE
Eliminate psimrcc's Memory Manager

### PR DESCRIPTION
**Pull request by JonathonMisiewicz**
## Description
As part of #1877, and the ongoing subproject to give `psimrcc` a wavefunction, the goal of this PR is to eliminate the various `allocate` and `release` calls in psimrcc so we can remove the MemoryManager global. They will be replaced with `std::vector`, smart pointers, or more direct creation of raw pointers where appropriate.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] Modernizing of `psimrcc` memory management.
- [x] Speed is fine!

## Questions
- [x] ~~Is there a better way to handle the `std::vector` replacements in `blas_diis.cc`? I'm not happy that I need to do a `reinterpret_cast`, but it's at least (marginally) better than C-style casts. Just focus on the third commit.~~ There is no better way.
## Checklist
- [x] All psimrcc tests pass as of the last commit

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Comment by JonathonMisiewicz** [comment]
This is just about as far as this PR is going to go. Fully removing memory_manager would require a way to store how much memory has been "spent." Keeping track of that without globals requires a wavefunction. Once the wavefunction is in place to track memory, the MemoryManager can be removed from psimrcc without much hassle.

Two issues before potentially merging this:
1. I want to do some profiling to make sure that none of my pointer -> std::vector replacements hit performance. `.data()` and passing const refs is enough, according to my very preliminary tests, but I'll do some more in-depth ones once I figure out how to use VTune.
2. Do we want this as a 1.4 target? We really don't need more PR dependencies before 1.4, given that the 1.4 paper is already published, but 1.4 is still not release-ready. If we don't, then let's hold off on the merge until after 1.4.

File: psi4/src/psi4/psimrcc/operation_contraction.cc [hunk 1]
```diff
@@ -331,511 +330,5 @@ void CCOperation::contract_in_core(double** A_matrix, double** B_matrix, double*
...
-      F_DGEMM("n","n",&n,&m,&k,&factor,&(C_matrix[0][0]),&n,&(B_matrix[0][0]),&cols_B,&beta,&(A_matrix[0][0]),&n);
-    }
-  }
-  // CASE D, best case scenario
-  //
-  //       -------------       ---------------
-  //      |             |     |               |
-  //   A[A_l][A_r] = B[B_i][B_c] 2@2 C[C_c][C_i]
-  //           |                        |
-  //            ------------------------
```
Comment (loriab) [inline psi4/src/psi4/psimrcc/operation_contraction.cc:822]: should any of these notes go into an attic repo instead of the ether?

Comment (JonathonMisiewicz) [reply psi4/src/psi4/psimrcc/operation_contraction.cc:822]: I see no reason for it. There are comparable notes for the actually-in-use `setup_contractions` function earlier in the file.

File: psi4/src/psi4/psimrcc/manybody.h [hunk 2]
```diff
@@ -115,18 +115,17 @@ class CCManyBody {
...
-    void deallocate_triples_denominators();
-
-    double*** d3_ooo;
-    double*** d3_ooO;
-    double*** d3_oOO;
-    double*** d3_OOO;
-    double*** d3_vvv;
-    double*** d3_vvV;
-    double*** d3_vVV;
-    double*** d3_VVV;
```
[the remark this answers is not in the document]
Comment (JonathonMisiewicz) [reply psi4/src/psi4/psimrcc/manybody.h:129]: I have enough to do without reforming `libdpd`, thank you. Though I'm not sure three nested vectors is much better...

**Comment by fevangelista** [comment]
Have you tried to use a `static_cast<char*>` instead of `reinterpret_cast<char*>`?

**Comment by JonathonMisiewicz** [comment]
Yes, I tried the static_cast first.

I get the error `static_cast from 'std::__1::vector<double, std::__1::allocator<double> >::value_type *' (aka 'double *') to 'char *' is not allowed`.

**Comment by fevangelista** [comment]
Yes, sorry, I think the `reinterpret_cast<char*>` here is necessary but it's really not much different than doing type conversion via (char*). In this case it's pretty harmless to do the cast since you are just sending the data to one function and not passing the pointer around or storing it somewhere else.

**Comment by hokru** [comment]
> 1. I want to do some profiling to make sure that none of my pointer -> std::vector replacements hit performance.

I  am in no way qualified to comment on the C++ pointer stuff. But I had a Intel VTune setup for this PR and [the profiling](https://github.com/psi4/psi4/files/4687387/result.txt) looks sufficiently clean.

**Comment by JonathonMisiewicz** [comment]
I made a few more changes after you ran that one. In particular, I want to check that [this commit](https://github.com/psi4/psi4/pull/1901/commits/e2539d75545fefbfc0e6f9306301b36b187cb5b3) didn't cause threading problems.

**Comment by hokru** [comment]
The above VTune output corresponds to  f5d93501075246ce6fc090f5f3b709a2fa956606 (current HEAD of this PR)

File: psi4/src/psi4/psimrcc/index.cc [hunk 3]
```diff
@@ -399,7 +362,7 @@ void CCIndex::make_three_index() {
     }
 
     // Allocate the memory for the tuples and store them
-    allocate2(short, tuples, ntuples, 3);
+    tuples = std::vector<std::vector<short>>(ntuples, std::vector<short>(3));
```
[the remark this answers is not in the document]
Comment (JonathonMisiewicz) [reply psi4/src/psi4/psimrcc/index.cc:365]: I considered that but couldn't find a way to make it work. Depending on whether you call `make_three_index`, `make_two_index`, or `make_one_index`, you'd want a `std::vector<std::array<short, 3>>` or `std::vector<std::array<short, 2>>` or `std::vector<std::array<short, 1>>`. A single variable can't be all three types.

File: psi4/src/psi4/psimrcc/index.h [hunk 4]
```diff
@@ -154,17 +153,17 @@ class CCIndex {
...
-    int* one_index_to_irrep;                   // 1->irrep mapping array
-    int** two_index_to_irrep;                  // 2->irrep mapping array
-    int*** three_index_to_irrep;               // 3->irrep mapping array
-    int** element_irrep;                       // Irrep of each element
+    Size_tVec one_index_to_tuple_rel_index;    // 1->tuple mapping array
+    std::vector<std::vector<size_t>> two_index_to_tuple_rel_index;     // 2->tuple mapping array
+    std::vector<std::vector<std::vector<size_t>>> three_index_to_tuple_rel_index;  // 3->tuple mapping array
+    std::vector<int> one_index_to_irrep;       // 1->irrep mapping array
+    std::vector<std::vector<int>> two_index_to_irrep;                  // 2->irrep mapping array
+    std::vector<std::vector<std::vector<int>>> three_index_to_irrep;               // 3->irrep mapping array
```
[the remark this answers is not in the document]
Comment (JonathonMisiewicz) [reply psi4/src/psi4/psimrcc/index.h:165]: I don't see how. We only know the size of these quantities at run-time, but both `std::array` and `std::tuple` need that information at compile-time to specify the type, right?

**Comment by JonathonMisiewicz** [comment]
From my perspective, we need to figure out the tuple data type in `index.cc` before I can flatten the `tuple` member of the `IndexIterator`, before I can run final speed tests.

I was hoping Jet could show me VTune by then, but if that doesn't happen, I'll just run some computations from Allen's recent paper using Mk-MRCC with both versions and see if the update causes notable slowdowns.

**Comment by fevangelista** [comment]
Why not just store std::array<short,3> for all types (1-3 indices)? It turns out it is more expensive to store an uninitialized vector than the array:
```
sizeof(std::array<short,3>) = 6
sizeof(std::vector<short>) = 24% 
```

File: psi4/src/psi4/psimrcc/index_iterator.h [hunk 5]
```diff
@@ -83,9 +83,11 @@ class CCIndexIterator {
     int current_block;
 
     // Properties of the tuples
+    // These come from the Index object. Which should outlast the iterator, and which should
+    // not be changing these values as long as this iterator is alive.
     int nelements;
-    int** element_irrep;
-    short** tuples;
+    std::vector<std::vector<int>> element_irrep;
+    const std::vector<std::vector<short>>& tuples;
```
[the remark this answers is not in the document]
Comment (JonathonMisiewicz) [reply psi4/src/psi4/psimrcc/index_iterator.h]: Done. There's still going to be copy overhead, though.

File: psi4/src/psi4/psimrcc/index.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (JonathonMisiewicz) [reply psi4/src/psi4/psimrcc/index.h:166]: Using `std::array<short, 3>` won't work here, as each vector should have the length of the MO space, which we have no bound on.

I _could_ make these into `std::map<tuple<int, int, int>, int>` and such if you expect large performance gains from it.

**Comment by JonathonMisiewicz** [comment]
I'm going to learn VTune and then start running timing tests. Let me know if there are any things you want me to be sure to test.

**Comment by JonathonMisiewicz** [comment]
Now that I can run some VTune tests myself, there is definitely an efficiency hit from this PR. The explicit copy in the last commit contributes a _lot_ of overhead. If we want to flatten the tuples, this should be done in the index itself, not in the `IndexIterator`. With that commit reverted, we're looking at 62.4 seconds (pre-PR) vs 75.5 seconds (post-PR) for an Mk-MRCCSD(T) single point with 82 basis functions divided into irreps of 53 and 29, in serial. I can give more detailed machine diagnostics if needed, but I think the point is clear.

I'll work on getting the speed restored while waiting for reviews on my other live PR. This is a good opportunity to learn about how to make C++ code efficient, at least.

**Comment by JonathonMisiewicz** [comment]
After removing the last commit, new timings on CCQC machines for a system with 4 threads, 122 A' basis functions, 80 A'' basis functions, Mk-MRCCSD(T) on a trivial one-determinant model space: 1040s averaged over 5 runs before this PR, 995s averaged over 4 runs after. So once I got rid of that last commit, this PR is a small net speed _gain_ for PSIMRCC. I can run any additional tests if requested. 


So I say it's time for a last review. (@andysim?)

I just need to know what "best practice" for the commit history is. Reset away the last commit and force push, or create a new commit explicitly reverting the previous one? My vote is for the first one.

**Comment by loriab** [comment]
> I just need to know what "best practice" for the commit history is. Reset away the last commit and force push, or create a new commit explicitly reverting the previous one? My vote is for the first one.

In my view, so long as PR branch is solely yours, you're at perfect liberty to git reset, interactive rebase, fixup, force push, squash, etc.

**Pull request merged**